### ISSUE TITLE
Add new cinema options for camera combos

### ISSUE THREAD

**Issue by cyrush** [Alpine-DAV/ascent#828]
1. User provides two lists of values (phi, theta) and the cross product of those lists are the camera positions output. For example:

    phi: [0,90]
    theta: [0,45]

    Would create camera views from [phi, theta] pairs of [0, 0] [0, 45] [90, 0] [90, 45]

    2. User provides a list of pairs, and these are the [phi, theta] pairs of camera positions output. For example:

    camera_positions: [[0,0], [0,10], [0,20], [0,30]]

    This would create four camera positions.

    3. User provides two sets of range, increment data, and the cross product of the camera positions are created. For example:

    phi: [0-90], 10
    theta: [0-90], 10

    would be shorthand for:

    phi: [0,10,20,30,40,50,60,70,80,90]
    theta: [0,10,20,30,40,50,60,70,80,90]

**Comment by cyrush** [comment]
@dhrogers

**Comment by cyrush** [comment]
@brugger1

**Comment by cyrush** [comment]
@brugger1  here are some links for context:

This is the current class the generates the cameras from current params:

https://github.com/Alpine-DAV/ascent/blob/56ecd3d43e67e596fa11b047898973297797af13/src/ascent/runtimes/flow_filters/ascent_runtime_rendering_filters.cpp#L491

Here is where we actually connect up the params passed from the user:

https://github.com/Alpine-DAV/ascent/blob/56ecd3d43e67e596fa11b047898973297797af13/src/ascent/runtimes/flow_filters/ascent_runtime_rendering_filters.cpp#L1033


Instead of just having `phi` and `theta` be pulled out and passed as integers, we can pass multiple variants via a conduit node and have CinemaManager expand them however we like.

**Comment by brugger1** [comment]
@cyrush Since there isn't anyway to distinguish between a range and increment (3 values) and a list of values, I believe I need to introduce new key words in addition to theta, phi. I can see 2 approaches:

1) Add new key words for the list of values and for the range. There might still be ambiguity if the user specifies more than one. Could define a precedence if more than one or add a keyword that indicates which one to use.
2) Add a new key word on how to interpret the existing phi and theta keywords.

For the first approach:

List of theta and phi angles:
```
phi_values
theta_values
```

Range of theta and phi angle with increment:
```
phi_range
theta_range
```

For the second approach:
```
angle_spec_type: num_angles, angle_values, angle_range
```

**Comment by cyrush** [comment]
We can leave the existing and try:
`phi_range`
`theta_range`
`num_angles`

One Q I have, do we want a independent increment for phi or theta? If so, is the cartestian product the final set of cameras we want? 

Options like:

`phi_range`
`phi_num_angles`
`theta_range`
`theta_num_angles`

`phi_range`
`dphi`
`theta_range`
`dtheta`

**Comment by brugger1** [comment]
I think we would want independent increments for phi and theta and have the final set of cameras be the cartesian product of them. I like the set with `dphi` and `dtheta`. It more closely matches the familiar looping construct in Python.

**Comment by cyrush** [comment]
With providing the range, the delta can be a awkward.
 
we might need both (people will some times know the delta, and some times know the number) for blueprint, here is how we handle this:

```
origin/x
spacing/dx
(number is inferred from logical dim)
```
this way the top end of the range is not needed.

**Comment by brugger1** [comment]
Ah, so support the following 4 combinations from what you described above
```
phi_range, phi_num_angles
theta_range, theta_num_angles
phi_range, dphi
theta_range, dtheta
```
Sounds good to me!

**Comment by brugger1** [comment]
While debugging my changes I was looking at the behavior of phi and theta as currently implemented.

phi range: -180 -> 180
theta range: 0 -> 180

Using the defaults:
phi = 2, theta = 2
gives
phi = -180, 0
theta = 0, 90

This seems a little odd, I would almost expect
phi = -180, 180
theta = 0, 180
That might not be very useful, since we end up looking from the bottom up and the top down, so I can maybe understand why it was done this way. Maybe that is why we are making this enhancement. I'm not proposing changing this, but just making note of it.

**Comment by brugger1** [comment]
Reading the Ascent documentation, it indicates that `phi` is the polar angle and `theta` is the azimuth angle. This is consistent with Wolfram.

https://mathworld.wolfram.com/PolarAngle.html
https://mathworld.wolfram.com/Azimuth.html

If this is the case then `phi` should be in the range 0 -> 180, and `theta` in the range -180 -> 180, which is backwards from the code. Also, if I look at the resulting cinema database, it looks like changing `phi` with `theta=0` or `theta=180` results in the image spinning around an axis in the center of the image, which to me implies that `phi` is the azimuth and `theta` the polar angle.

**Comment by dhrogers** [comment]
Hey. Chiming into the discussion.

In addition to the range options you are discussing, the list option (#2 in the first message in this chain) is important, as it allows a set of arbitrary values. It's not that compact, but it can represent any set of viewpoints.

**Comment by brugger1** [comment]
For #2 above, how about `camera positions` with a list of phi, theta combinations. For example:
```
camera_positions = [0, 0, 0, 10, 0, 20]
```
It seems like a list of pairs of angles is overkill.

For #1 above, how about `phi_angles` and `theta_angles`? For example:
```
phi_angles = [0, 45]
theta_angles = [0, 45]
```

**Comment by dhrogers** [comment]
Will your `camera_positions` list above resolve to the following three `[phi, theta]` positions?
```
[0,0] [0,10] [0,20]
```
It's up to you, if you want a single list or a list of lists. I would like the ability to specify a collection of arbitrary positions that ascent can make sense of. My feedback on the single list is that it's not very human readable, but I can live with it either way.

It might be best to use the keyword `phi_theta_positions` or something like that instead of `camera_positions`, to stay consistent with the other names, and specifically call out that the values are angles.

`phi_angles` and `theta_angles` for #1 seems good.

**Comment by brugger1** [comment]
I agree `phi_theta_positions` would be a better choice for the name.

Regarding the single list versus an array of tuples. I don't have a good understanding on how difficult it is to handle multi-dimensional arrays in Conduit. @cyrush do you have any thoughts on this?

**Comment by brugger1** [comment]
I believe implicit in the cinema A spec viewers is the concept that there are a list of `phi` angles and a list of `theta` angles and that the total angles forms a cross product of the 2. For example, when I bring up the generic viewer there is a slider for `phi` and a slider for `theta`. What if the phi and theta values don't form a cross product? Does having a list of arbitrary `phi` and `theta` values make sense?

**Comment by brugger1** [comment]
@dhrogers is there a concept of a standard for the `phi` and `theta` angles? Does it follow the conventions at Wolfram

https://mathworld.wolfram.com/PolarAngle.html
https://mathworld.wolfram.com/Azimuth.html

Where `phi` is the polar angle down from the positive Z axis, going from 0 -> 180 and `theta` is an azimuthal angle in the X-Y plane from the positive X axis going from 0 -> 360?

**Comment by brugger1** [comment]
#1 and #3 are complete. I'm waiting on input from @dhrogers before completing #2. If forgot to specifically tag David on my question 2 posts up in this thread.

**Comment by dhrogers** [comment]
@brugger1 - 

SIMPLE ANSWER: The specific definition of `phi` and `theta` aren't constrained by the cinema specification. Instead, it is expected that the producer and consumer of the database know what those values represent.

In general, this means the rotations are relative to the meaning of the data. Since the data is generally computed in the `physics` domain, the wolfram makes sense as a default.

Cinema doesn't care about the 0->180 and 0->360 constraints, because everything is a range of values. A value of 720 would be OK in the Cinema database. It's up to the producer to constrain that.

MORE COMPLEX ANSWER: In general, the simulations don't have a standard for XYZ orientation, and even if they did, a user might want to rotate a specific orientation of the data with `phi` and `theta` controls. I hadn't thought about this before, but this brings up the point that a general solution API for a capability like Ascent should include the following parameters (not necessary for this implementation, but it would make the API general):

`phi-axis` the axis around which the `phi` angle rotates
`theta-axis` the axis around which he `theta` angle rotates
`axis-center` the center point for the phi/theta rotation
`camera-distance` the distance from the rotation center that the camera is

For this Milestone, it would be great, but not necessary, to implement these parameters. We can add them to a feature request if that works better.

**Comment by brugger1** [comment]
@dhrogers Those are great options to add. I may do that in another PR.

Currently, part of the cinema database includes a `data.csv` file, which can be left alone for the `phi_theta_positions`, but I believe we would need to remove the `phi` and `theta` sections from the `info.json` file since there would no longer potentially be a cross product of angles. Which I believe would make the file unusable by the current viewer. Does that sound correct?

**Comment by dhrogers** [comment]
`info.json` is no longer needed.

**Comment by brugger1** [comment]
Awesome!

**Comment by brugger1** [comment]
I added `phi_theta_positions` and dropped `phi` and `theta` from `info.json` and it broke the cinema viewer that Ascent outputs with the cinema database. The images generated are correct. Probably not critical, but something to note in the documentation since someone wanting to look at the output with the Ascent viewer probably wouldn't use `phi_theta_positions`.

**Comment by dhrogers** [comment]
Clarification: the `info.json` file is no longer needed, and can be entirely eliminated. The spec calls for `data.csv` and associated files only.

**Comment by cyrush** [comment]
@brugger1 since folks are using the old viewer (that's what Ascent produces for them) , we might still want `info.json` for cases that it is possible to keep them.

**Comment by cyrush** [comment]
> I agree `phi_theta_positions` would be a better choice for the name.
> 
> Regarding the single list versus an array of tuples. I don't have a good understanding on how difficult it is to handle multi-dimensional arrays in Conduit. @cyrush do you have any thoughts on this?

We could use a list of arrays:

[ [0,0,0], [z,z,z], ... ]

```
>>> n.parse("[ [0,0,0], [1,1,1] ]","yaml")
>>> n

- [0, 0, 0]
- [1, 1, 1]
```

That would show up in conduit as a `list` with leaves that are numeric arrays.

You can check for that case by asking if the data type is a list.

**Comment by dhrogers** [comment]
Might be better to update to the newer viewer, if that's possible. I can help with how to install that automatically.

**Comment by cyrush** [comment]
@dhrogers I would like to update. Ideally, we need the no-xhr edits (that I have currently hacked in) to  allow us to generate viewers that can run with out browser security tweaks.

**Comment by brugger1** [comment]
@cyrush It still generates the `info.json` file and if `phi_theta_positions` is used, the the `phi` and `theta` values are missing from the file.

@cyrush How would I do a list of arrays in C?

@dhrogers I'm definitely interested in the new viewer. Where do I find it?

**Comment by cyrush** [comment]
@brugger1 

```
conduit::Node n;
conduit::Node & list_entry = n.append();
list_entry.set({0.0,0.0,0.});

// each append() gives you a reference to new node appended to the list

```

**Comment by cyrush** [comment]
```
// to access the values, you can use a NodeIterator, or

float64 *ptr = n[0].value();
```

**Comment by dhrogers** [comment]
@brugger1 Repository is here: https://github.com/cinemascience/cinema_viewers 

The simplest viewer is `cinema_view.html`, which is expecting the `cinema` directory as well. I should take a pass at simplifying the installation, and making sure things point to local resources (they can also point to web URLs).

You will have to edit the `cinema/view/2.0/databases.json` file to point to your database, and the files should be structure like this:

```
somedir/
    results.cdb
    cinema_view.html
    cinema/
```

Not sure if these instructions are in the doc/readme, but I can help if it's not clear.

**Comment by dhrogers** [comment]
Note that a single viewer can point to a number of databases, if the data is arranged like this:

```
somedir/
    result001.cdb/
    result002.cdb/
    result003.cdb/
    cinema/
    cinema_view.html
```

Depending on the contents of the `databases.json` file, you can view one or several databases at a time. You can also have several different views of the data available from the dropdown menu.
